### ISSUE TITLE
Improve foreign key selector when dealing with lots of rows

### ISSUE THREAD

**Issue by dantownsend** [piccolo-orm/piccolo_admin#30]
The foreign key selectors are `select` HTML elements. These are used in the sidebar for filtering, and for changing foreign key values in detail views.

It works fine when the number of values is relatively small. If there are thousands of values, it could become less useful.

<img width="320" alt="Screenshot 2021-02-03 at 20 12 22" src="https://user-images.githubusercontent.com/350976/106803444-22b14780-665c-11eb-8389-bf7406f4f0ba.png">

A search box might be better in these situations.

`PiccoloCRUD` has an `ids` endpoint - `/api/tables/<tablename>/ids` which returns a mapping of row ID to a readable value. For example:

```json
{
    "1": "Peter Jackson",
    "2": "George Lucas",
    "3": "J.J. Abrams",
    "4": "Gareth Edwards",
    "5": "Irvin Kershner",
    "6": "Richard Marquand",
    "7": "Rian Johnson",
    "8": "Ron Howard"
}
```
Try it out [here](https://demo1.piccolo-orm.com/api/tables/director/ids/).

The endpoint could be modified to return data in this format:

```json
{
    "total_count":  8,
    "response_count": 8,
    "rows": {
        "1": "Peter Jackson",
        "2": "George Lucas",
        "3": "J.J. Abrams",
        "4": "Gareth Edwards",
        "5": "Irvin Kershner",
        "6": "Richard Marquand",
        "7": "Rian Johnson",
        "8": "Ron Howard"
    }
}
```

If there were lots of values, it would return something like this:

```json
{
    "total_count":  5000,
    "response_count": 100,
    "rows": {
        "1": "Peter Jackson",
        "2": "George Lucas",
        "3": "J.J. Abrams",
        "4": "Gareth Edwards",
        "5": "Irvin Kershner",
        "6": "Richard Marquand",
        "7": "Rian Johnson",
        ...
        "100": "Quentin Tarantino",
    }
}
```
By default, a maximum of 100 results will be returned at once. If `response_count < total_count` we should show the search UI rather than just a dropdown.

The ids endpoint should accept a search parameter, which can be used to filter what's returned. For example:

```json
// /api/tables/director/ids?search=george
{
    "total_count":  2,
    "response_count": 2,
    "rows": {
        "1": "George Lucas",
        "50": "George Armitage",
    }
}
```
The search UI could look something like this:

![sZFsT](https://user-images.githubusercontent.com/350976/106806114-73766f80-665f-11eb-92ae-613907a0c98b.png)

**Comment by dantownsend** [comment]
Rather than modifying the `ids` endpoint - it might be better to keep that as is for backwards compatibility, but to implement some kind of cap (say 1000). The new endpoint could be called something like `id-search`.

**Comment by sinisaos** [comment]
@dantownsend Like I wrote yesterday this is my attempt on foreign key selection.

```
KeySelect.vue

<template>
    <div>
        <input
            :placeholder="'Search ' + tableName + 's'"
            autocomplete="off"
            type="text"
            v-bind:name="fieldName"
            v-model="selectedValue"
            v-on:focus="resetResult(), toggle = true"
            v-on:keydown.enter.prevent
        />
        <div
            id="results"
            v-if="toggle"
        >
            <div
                :key="id"
                v-for="(readable, id) in dropdownItems"
                v-on:click="selectResult(readable, id), toggle = false"
                v-on:focus="resetResult(), toggle = true"
            >{{ readable }}</div>
        </div>
        <input
            :value="hiddenSelectedValue"
            id="hidden-value"
            type="text"
            v-bind:name="fieldName"
            v-on:focus="toggle = true"
        />
    </div>
</template>

<script>
export default {
    props: {
        fieldName: String,
        tableName: String,
        value: undefined,
    },
    data() {
        return {
            ids: [],
            dropdownItems: {},
            selectedValue: undefined,
            hiddenSelectedValue: undefined,
            toggle: false,
        }
    },
    methods: {
        async fetchData() {
            const response = await this.$store.dispatch(
                "fetchIds",
                this.tableName
            )
            this.ids = response.data
        },
        selectResult(readable, id) {
            this.selectedValue = readable
            this.hiddenSelectedValue = id
        },
        // need to explicit clear values
        resetResult() {
            this.selectedValue = undefined
            this.hiddenSelectedValue = undefined
        },
    },
    watch: {
        async tableName() {
            await this.fetchData()
        },
        selectedValue(searchValue) {
            this.dropdownItems = {}
            for (const property in this.ids) {
                if (
                    this.ids[property]
                        .toLowerCase()
                        .includes(searchValue.toLowerCase())
                ) {
                    this.dropdownItems[property] = this.ids[property]
                }
            }
        },
    },
    async mounted() {
        this.selectedValue = this.value
        this.hiddenSelectedValue = this.value
        await this.fetchData()
    },
}
</script>


<style lang="less" scoped>
@import "../vars.less";

.light_mode {
    background-color: white;
    color: @dark_grey;
    #hidden-value {
        display: none;
    }
    #results {
        cursor: pointer;
        background-color: darken(white, 5%);
        color: @dark_grey;
    }
}

.dark_mode {
    background-color: @dark_grey;
    color: @off_white;
    #hidden-value {
        display: none;
    }
    #results {
        cursor: pointer;
        background-color: darken(@dark_grey, 4%);
        color: @off_white;
    }
}
</style>
```
and `RowForm.vue`

```
<template>
    <div>
        <div
            v-bind:key="property.title"
            v-for="(property, keyName) in schema.properties"
        >
            <template v-if="property.extra.foreign_key">
                <label>
                    {{ property.title }}
                    <span
                        class="required"
                        v-if="isRequired(keyName)"
                    >*</span>

                    <router-link
                        :to="{name: 'addRow', params: {tableName: property.extra.to}}"
                        class="add"
                        target="_blank"
                        v-if="!isFilter"
                    >
                        <font-awesome-icon icon="plus" />
                    </router-link>

                    <router-link
                        :to="{name: 'editRow', params: {tableName: property.extra.to, rowID: getKeySelectID(property.title)}}"
                        class="add"
                        target="_blank"
                        v-if="!isFilter"
                    >
                        <font-awesome-icon icon="edit" />
                    </router-link>
                </label>
                <KeySelect
                    v-bind:fieldName="property.title.toLowerCase()"
                    v-bind:key="getValue(property.title)"
                    v-bind:tableName="property.extra.to"
                    v-bind:value="getValue(property.title)"
                />
            </template>
            <template v-else>
                <label>
                    {{ property.title }}
                    <span
                        class="required"
                        v-if="isRequired(keyName)"
                    >*</span>
                </label>
                <InputField
                    v-bind:format="property.format"
                    v-bind:isFilter="isFilter"
                    v-bind:isNullable="property.nullable"
                    v-bind:key="property.title"
                    v-bind:required="isRequired(keyName)"
                    v-bind:title="property.title"
                    v-bind:type="property.type || property.anyOf[0].type"
                    v-bind:value="getValue(property.title)"
                />
            </template>
        </div>
    </div>
</template>

<script lang="ts">
import Vue from "vue"

import KeySelect from "./KeySelect.vue"
import InputField from "./InputField.vue"

export default Vue.extend({
    props: {
        row: Object,
        schema: Object,
        isFilter: {
            type: Boolean,
            default: false,
        },
    },
    components: {
        InputField,
        KeySelect,
    },
    data() {
        return {
            keySelectIDs: {},
        }
    },
    methods: {
        getValue(propertyTitle: string) {
            let value = this.row
                ? this.row[propertyTitle.toLowerCase().split(" ").join("_")]
                : undefined
            return value
        },
        getKeySelectID(propertyTitle: string) {
            return this.getValue(propertyTitle)
        },
        isRequired(keyName: string) {
            return (
                !this.isFilter &&
                (this.schema.required || []).indexOf(keyName) != -1
            )
        },
    },
})
</script>

<style scoped lang="less">
.add {
    float: right;
}

span.required {
    opacity: 0.5;
    padding-left: 0.05rem;
}
</style>
```
The biggest problem is that I can't get FK readable (`1` is displayed instead of `Peter Jackson`) in edit form. If you like it i can make PR or you can simply copy/paste and try it. Cheers.

**Comment by dantownsend** [comment]
@sinisaos Sorry for the slow response.

This looks good - thanks. If you create a pull request, I'll play around with it, and will try and work out the issue with the FK readable.

**Comment by dantownsend** [comment]
Fixed in version 0.11.10
